File: src/main/java/forpleuvoir/hiirosakura/client/gui/qcms/QCMSScreen.java
package forpleuvoir.hiirosakura.client.gui.qcms;

import fi.dy.masa.malilib.gui.*;
import fi.dy.masa.malilib.gui.button.ButtonBase;
import fi.dy.masa.malilib.gui.button.IButtonActionListener;
import fi.dy.masa.malilib.interfaces.IConfirmationListener;
import fi.dy.masa.malilib.util.StringUtils;
import forpleuvoir.hiirosakura.client.HiiroSakuraClient;
import forpleuvoir.hiirosakura.client.config.HiiroSakuraDatas;
import forpleuvoir.hiirosakura.client.feature.qcms.QuickChatMessage;
import forpleuvoir.hiirosakura.client.feature.qcms.QuickChatMessageSend;
import forpleuvoir.hiirosakura.client.feature.task.TimeTask;
import forpleuvoir.hiirosakura.client.feature.task.TimeTaskHandler;
import forpleuvoir.hiirosakura.client.util.Colors;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.LiteralText;
import net.minecraft.text.MutableText;
import net.minecraft.text.TranslatableText;

import java.util.LinkedList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * 快速聊天消息发送屏幕
 *
 * @author forpleuvoir
 * <p>#project_name hiirosakura
 * <p>#package forpleuvoir.hiirosakura.client.gui
 * <p>#class_name QCMSScreen
 * <p>#create_time 2021/6/18 21:12
 */
public class QCMSScreen extends GuiBase {
    private final TranslatableText title = new TranslatableText(
            String.format("%s.gui.title.qcms", HiiroSakuraClient.MOD_ID));
    private final TranslatableText empty = new TranslatableText(
            String.format("%s.feature.qcms.data.empty", HiiroSakuraClient.MOD_ID));
    private final List<QCMSButton> buttons = new LinkedList<>();

    public QCMSScreen() {
        super();
        this.setTitle(StringUtils.translate(title.getKey()));
    }

    @Override
    public void initGui() {
        super.initGui();
        var datas = HiiroSakuraDatas.QUICK_CHAT_MESSAGE_SORT.getSortedData();
        HiiroSakuraDatas.QUICK_CHAT_MESSAGE_SORT.getUnSortedData().forEach(datas::addLast);
        this.addButton(
                new QCMSButton(this.width / 2 - 40, this.height - 55, 80, 20,
                        StringUtils.translate(String.format("%s.gui.button.add", HiiroSakuraClient.MOD_ID))
                ),
                (button, mouseButton) -> openAddScreen()
        );
        if (datas.isEmpty()) {
            this.addGuiMessage(Message.MessageType.WARNING, 2000, empty.getKey());
            return;
        }
        int padding = 10;
        AtomicInteger indexX = new AtomicInteger(40);
        AtomicInteger indexY = new AtomicInteger(40);
        datas.forEach(qcm -> {
            int width = Math.max(QuickChatMessageSend.getKeyLength(qcm.remark()) * 12, 20);
            int x = indexX.get();
            indexX.addAndGet(width);
            int y = indexY.get();
            if (this.width - indexX.get() <= 40 || indexX.get() + width > (this.width - 20)) {
                indexY.addAndGet(20);
                indexX.set(40);
            }
            MutableText message = new LiteralText(qcm.message()).styled(style -> style.withColor(
                    Colors.DHWUIA.getColor()));
            MutableText tooltip = new TranslatableText(String.format("%s.gui.qcms.hover",
                    HiiroSakuraClient.MOD_ID
            )).styled(style -> style.withColor(Colors.FORPLEUVOIR.getColor()));
            var qcmsButton = new QCMSButton(x + padding, y + padding, width, 20,
                    qcm.remark().replace("&", "§"),
                    message, tooltip
            );
            this.addButton(qcmsButton, (button, mouseButton) ->
                    this.buttonClick(mouseButton, qcm)
            );

        });
    }

    private void sendChatMessage(String message) {
        assert Objects.requireNonNull(client).player != null;
        if (client.player != null) {
            client.player.sendChatMessage(message);
            TimeTaskHandler.getInstance()
                    .addTask(TimeTask.once(hiiroSakuraClient -> this.onClose(), 5, "Close_QCMS_Screen"));
        }
    }

    protected void drawButtonHoverTexts(int mouseX, int mouseY, float partialTicks, MatrixStack matrixStack) {
        for (QCMSButton button : buttons) {
            if (button.hasHoverText() && button.isMouseOver()) {
                renderTooltip(matrixStack, button.getHoverText(), mouseX + 10, mouseY);
            }
        }
    }

    @Override
    public <T extends ButtonBase> T addButton(T button, IButtonActionListener listener) {
        if (button instanceof QCMSButton qcmsButton) {
            this.buttons.add(qcmsButton);
        }
        return super.addButton(button, listener);
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }

    private void buttonClick(int mouseButton, QuickChatMessage quickChatMessage) {
        switch (mouseButton) {
            case 0 -> sendChatMessage(quickChatMessage.message());
            case 1 -> openEditScreen(quickChatMessage);
            case 2 -> openDeleteScreen(quickChatMessage.remark());
        }
    }

    private void openDeleteScreen(String key) {
        String titleKey = String.format("%s.gui.title.qcms.delete", HiiroSakuraClient.MOD_ID);
        String messageKey = String.format("%s.gui.qcms.confirmDelete", HiiroSakuraClient.MOD_ID);
        int width = Math.max(textRenderer
                        .getWidth(StringUtils.translate(titleKey)),
                textRenderer
                        .getWidth(StringUtils.translate(messageKey))
        );
        var dialog = new GuiConfirmAction(width,
                titleKey,
                new IConfirmationListener() {
                    @Override
                    public boolean onActionConfirmed() {
                        HiiroSakuraDatas.QUICK_CHAT_MESSAGE_SEND.remove(key);
                        HiiroSakuraDatas.QUICK_CHAT_MESSAGE_SORT.remove(key);
                        return true;
                    }

                    @Override
                    public boolean onActionCancelled() {
                        return false;
                    }
                },
                this,
                messageKey,
                key
        );
        openGui(dialog);
    }

    private void openEditScreen(QuickChatMessage quickChatMessage) {
        GuiBase.openGui(new EditScreen(quickChatMessage, this));
    }

    private void openAddScreen() {
        GuiBase.openGui(new EditScreen(this));
    }


}

File: src/main/kotlin/forpleuvoir/hiirosakura/client/gui/qcms/QCMSScreen.kt
package forpleuvoir.hiirosakura.client.gui.qcms

import fi.dy.masa.malilib.gui.GuiBase
import fi.dy.masa.malilib.gui.GuiConfirmAction
import fi.dy.masa.malilib.gui.Message
import fi.dy.masa.malilib.gui.button.ButtonBase
import fi.dy.masa.malilib.gui.button.IButtonActionListener
import fi.dy.masa.malilib.interfaces.IConfirmationListener
import fi.dy.masa.malilib.util.StringUtils
import forpleuvoir.hiirosakura.client.HiiroSakuraClient
import forpleuvoir.hiirosakura.client.config.HiiroSakuraData
import forpleuvoir.hiirosakura.client.feature.qcms.QuickChatMessage
import forpleuvoir.hiirosakura.client.feature.task.TimeTask.Companion.once
import forpleuvoir.hiirosakura.client.feature.task.TimeTaskHandler.Companion.INSTANCE
import forpleuvoir.hiirosakura.client.util.Colors
import forpleuvoir.hiirosakura.client.util.StringUtil.translatableText
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.LiteralText
import net.minecraft.text.Style
import net.minecraft.text.TranslatableText
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer


/**
 * 快捷聊天消息发送GUI

 * 项目名 hiirosakura

 * 包名 forpleuvoir.hiirosakura.client.gui.qcms

 * 文件名 QCMSScreen

 * 创建时间 2021-08-18 16:02

 * @author forpleuvoir

 */
@Deprecated("准备删除")
class QCMSScreen : GuiBase() {
	private val empty = TranslatableText(String.format("%s.feature.qcms.data.empty", HiiroSakuraClient.MOD_ID))
	private val buttons: LinkedList<QCMSButton> = LinkedList()

	init {
		setTitle(StringUtils.translate(translatableText("gui.title.qcms").key))
	}

	override fun initGui() {
		super.initGui()
		val data = HiiroSakuraData.QUICK_CHAT_MESSAGE_SORT.getSortedData(true)
		HiiroSakuraData.QUICK_CHAT_MESSAGE_SORT.unSortedData.forEach(Consumer { e: QuickChatMessage ->
			data.addLast(
				e
			)
		})
		addButton(
			QCMSButton(
				width / 2 - 40, height - 55, 80, 20,
				StringUtils.translate(String.format("%s.gui.button.add", HiiroSakuraClient.MOD_ID))
			)
		) { _: ButtonBase?, _: Int -> openAddScreen() }
		if (data.isEmpty()) {
			addGuiMessage(Message.MessageType.WARNING, 2000, empty.key)
			return
		}
		val padding = 10
		val indexX = AtomicInteger(40)
		val indexY = AtomicInteger(40)
		data.forEach(Consumer { qcm: QuickChatMessage ->
			val remark = qcm.remark.replace("&", "§")
			val width = (this.mc.textRenderer.getWidth(remark) + 12).coerceAtLeast(20)
			val x = indexX.get()
			indexX.addAndGet(width + 2)
			val y = indexY.get()
			if (this.width - indexX.get() <= 40 || indexX.get() + width > this.width - 20) {
				indexY.addAndGet(20)
				indexX.set(40)
			}
			val message =
				LiteralText(qcm.message).styled { style: Style ->
					style.withColor(
						Colors.DHWUIA.color
					)
				}
			val tooltip = TranslatableText(
				String.format(
					"%s.gui.qcms.hover",
					HiiroSakuraClient.MOD_ID
				)
			).styled { style: Style ->
				style.withColor(
					Colors.FORPLEUVOIR.color
				)
			}
			val qcmsButton = QCMSButton(
				x + padding, y + padding, width, 20,
				remark,
				message, tooltip
			)
			addButton(qcmsButton,
				IButtonActionListener { _: ButtonBase?, mouseButton: Int -> this.buttonClick(mouseButton, qcm) }
			)
		})
	}

	private fun sendChatMessage(message: String) {
		HiiroSakuraClient.sendMessage(message)
		INSTANCE!!.addTask(
			once(
				5,
				"#Close_QCMS_Screen"
			) {
				onClose()
			}
		)
	}

	override fun drawButtonHoverTexts(mouseX: Int, mouseY: Int, partialTicks: Float, matrixStack: MatrixStack?) {
		for (button in buttons) {
			if (button.hasHoverText() && button.isMouseOver) {
				renderTooltip(matrixStack, button.getHoverText(), mouseX + 10, mouseY)
			}
		}
	}

	override fun <T : ButtonBase> addButton(button: T, listener: IButtonActionListener): T {
		if (button is QCMSButton) {
			this.buttons.add(button)
		}
		return super.addButton(button, listener)
	}

	override fun isPauseScreen(): Boolean {
		return false
	}

	private fun buttonClick(mouseButton: Int, quickChatMessage: QuickChatMessage) {
		when (mouseButton) {
			0 -> sendChatMessage(quickChatMessage.message)
			1 -> openEditScreen(quickChatMessage)
			2 -> openDeleteScreen(quickChatMessage.remark)
		}

	}

	private fun openDeleteScreen(key: String) {
		val titleKey = String.format("%s.gui.title.qcms.delete", HiiroSakuraClient.MOD_ID)
		val messageKey = String.format("%s.gui.qcms.confirmDelete", HiiroSakuraClient.MOD_ID)
		val width = textRenderer
			.getWidth(StringUtils.translate(titleKey)).coerceAtLeast(
				textRenderer
					.getWidth(StringUtils.translate(messageKey))
			)
		val dialog = GuiConfirmAction(
			width,
			titleKey,
			object : IConfirmationListener {
				override fun onActionConfirmed(): Boolean {
					HiiroSakuraData.QUICK_CHAT_MESSAGE_SEND.remove(key)
					HiiroSakuraData.QUICK_CHAT_MESSAGE_SORT.remove(key)
					return true
				}

				override fun onActionCancelled(): Boolean {
					return false
				}

			},
			this,
			messageKey,
			key
		)
		openGui(dialog)
	}

	private fun openEditScreen(quickChatMessage: QuickChatMessage) {
		openGui(EditScreen(quickChatMessage, this))
	}

	private fun openAddScreen() {
		openGui(EditScreen(this))
	}


}